### ISSUE TITLE
oAuth2 with implict flow for OpenID

### ISSUE THREAD

**Issue by dschneid85** [xamarin/Xamarin.Auth#228]
Hello,

I use your oauth framework to authenticate with our own oauth server.
This server is an connect2id server which implement the openid standard.
We would like to use the oauth2 implict flow with your framework but this isn't possible.
The connect2id server required for the implict flow the response_type=id_token or response_type=id_token%20token.
Therefore the nonce parameter is also required. So I have to set this parameter if I use the implict flow.
Check this in the open id standard https://openid.net/specs/openid-connect-core-1_0.html#ImplicitFlowAuth. I think your oAuth framework should be worked according to the open id standard.
What I have did I checked out your github project and changed a bit the class OAuth2Authenticator.cs.
I added the nonce parameter and changed the response_type to id_token token. Then your code works fine for me.
Another problem what I have is how can I change the title page of the embedded webview. The title can I changed (authenticator.Title ) but not the icon and the background color.
Is there any way to change this?

Best regards,

Daniel

**Comment by moljac** [comment]
@dschneid85 
Xamarin.Auth was designed to support OAuth and from OAuth standpoint OpenID is superset/extension of OAuth.
There are some parts in the API using IDictionary<string, string> for parameters, but I haven't gone that far - I need to implement/fix few more OAuth features before going in the direction of "custom parameters" for OpenId. To give you few good news - I'm working on that currently together with some other features (in parallel). One of those features/improvements is `response_type` which is currently hardcoded (as you probably know). Sorry, but I have to be very careful not to break Xamarin.Auth for other people.

Can you share changes to the OAuth2Authenticator, so I can investigate and possibly speed up a bit?

> change the title page of the embedded webview. The title can I changed (authenticator.Title ) but not the icon and the background color.

I think there is an issue for that. I need to go deeper into the code for that. I know Title can be changed, but I'm quite sure I haven't seen Icon and color API. BTW which icon and on which platform?

**Comment by dschneid85** [comment]
Hello moljac,

thanks for your fast answer. Sure, I attached my changes as txt file and the instantiation part as picture.
 So you can see how I changed your code to use implicit flow accordingly to the OpenId standard.

![instantiationoauth2authenticator](https://user-images.githubusercontent.com/31453219/30415411-3c277d74-9928-11e7-9728-fc60ea32a57f.png)

[OAuth2Authenticator.cs.txt](https://github.com/xamarin/Xamarin.Auth/files/1301942/OAuth2Authenticator.cs.txt)

I hope you can support the OpenID subset of the OAuth standard in the feature. 

I would like to set our own company icon in the toolbar and support the platforms Android and iOS.
May be it's possible to set the icon and the background color with a ControlTemplate to theme the embedded webview like the example https://developer.xamarin.com/samples/xamarin-forms/Templates/ControlTemplates/SimpleTheme/.

**Comment by moljac** [comment]
@dschneid85 

So you basically added 2 properties? Makes sense.

I'm basically working right now in `OAuthFlowResponseTypeVerification()` because initial Xamarin.Auth covered (was designed) mainly for ImplicitFLow and now we cover Authorization Code grant, PKCE flow, and other... 
I just need to go through the forest of ifs and other spaghetti to see where I can add other properties without breaking too much...

My design currently is going in the direction of adding Additional Request Paremeters (Nonce is parameter additional required parameter on top of OAuth `response_type` and `redirect_url`).

You hard coded generation of nonce parameter value, but I was thinking to add Func<> (delegate) and provide default implementation (like yours - I might use yours) and leave to the user the possibility to add their implementation of nonce value (maybe they need something stronger for random string - your is 16 chars)

Which OpenID provider are you using?
I have minimal list of free/public OpenId providers for testing (Google (Identity platform), Microsoft, Yahoo, PayPal, Amazon, SalesForce) but can you provide some more, please?

**Comment by dschneid85** [comment]
I thought that you would expanded your
OAuthFlowResponseTypeVerification() method
for the feature.
I needed a fast workaround to work with your code. Maybe it is better to
implement a base interface that supports different oauth flows. Then you
can implement different oauth flow classes which supports the required and
optional parameters. Maybe that is one way to eliminate the spaghetti code.

You can absolutely use my implementation of the nonce parameter.

We have an own OpenID server for our infrastructure. We use the connect2id
server to provide our own oauth service. Connect2id (https://connect2id.com/)
is an implementation of the OpenID standard.

You can use facebook and github as oAuth services for more testing your
code.

Best regards


14.09.2017 10:06 schrieb "Miljenko Cvjetko moljac moki mel" <
notifications@github.com>:

@dschneid85 <https://github.com/dschneid85>

So you basically added 2 properties? Makes sense.

I'm basically working right now in OAuthFlowResponseTypeVerification()
because initial Xamarin.Auth covered (was designed) mainly for ImplicitFLow
and now we cover Authorization Code grant, PKCE flow, and other...
I just need to go through the forest of ifs and other spaghetti to see
where I can add other properties without breaking too much...

My design currently is going in the direction of adding Additional Request
Paremeters (Nonce is parameter additional required parameter on top of
OAuth response_type and redirect_url).

You hard coded generation of nonce parameter value, but I was thinking to
add Func<> (delegate) and provide default implementation (like yours - I
might use yours) and leave to the user the possibility to add their
implementation of nonce value (maybe they need something stronger for
random string - your is 16 chars)

Which OpenID provider are you using?
I have minimal list of free/public OpenId providers for testing (Google
(Identity platform), Microsoft, Yahoo, PayPal, Amazon, SalesForce) but can
you provide some more, please?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/xamarin/Xamarin.Auth/issues/228#issuecomment-329406606>,
or mute the thread

**Comment by newky2k** [comment]
no longer needed

**Comment by rogerskk** [comment]
I know this is closed, but  I am trying to get Oauth2Authenticator working with WSO2 and need the id_token returned.  I tried using VerifyOAuth2FlowResponseType and it appears to send response_type=id_token in the request, but I don't get the id_token back in the response, only the access_token.  Any suggestions?
